### ISSUE TITLE
Make rule_haskell amenable to the persistent worker strategy

### ISSUE THREAD

**Issue by ulysses4ever** [tweag/rules_haskell#954]
**Describe the solution you'd like**
I want `rule_haskell` to make Bazel call GHC in the persistent worker mode, as it is the case [for the Java compiler](https://blog.bazel.build/2015/12/10/java-workers.html).

**Comment by ulysses4ever** [comment]
I was experimenting with making `rules_haskell` to run GHC in a (supposedly added afterward) daemon mode. This is called "persistent worker strategy" in Bazel-speak. That would mean, I believe, Bazel passing the `--persistent-worker` flag to GHC on the first occurrence of the `run_ghc` action during the build, and sending messages to the spawned GHC in the special protobuf-based format thereafter. Here are a couple of elaborating links if you're curious:

1. [Post on the Bazel Blog: Persistent Worker Processes for Bazel][bazelblog]

2. [Post on medium.com: How to Create a Persistent Worker for Bazel][medium]

[bazelblog]: https://blog.bazel.build/2015/12/10/java-workers.html
[medium]: https://medium.com/@mmorearty/how-to-create-a-persistent-worker-for-bazel-7738bba2cabb

Following these resources, in order to make Bazel use the persistent worker mode, one needs two things:

1) call `bazel build` with `--strategy=<mnemonic>=woker`. In the case of `rules_haskell`, I believe `<mnemonic>` is either `HaskellBuildLibrary` or `HaskellBuildExecutable` (there are variants for profiling and dynamic library modes though). This will do for initial experiments, I guess.

2) the `run` / `run_shell` action should  get `execution_requirements = { "supports-workers": "1" }` parameter. I believe the `run_shell` I'm interested in is in `toolchain.bzl`, in the implementation of `run_ghc` rule. The problem I'm facing now is after adding the parameter to this call to `run_shell`, `bazel build` started to complain (example when building first tutorial example):

    ```text
    $ bazel build --strategy=HaskellBuildLibrary=worker @io_tweag_rules_haskell_tutorial//lib:booleans
    Starting local Bazel server and connecting to it...
    INFO: Analysed target @io_tweag_rules_haskell_tutorial//lib:booleans (30 packages loaded, 609 targets configured).
    INFO: Found 1 target...
    ERROR: 
    /home/artem/.cache/bazel/_bazel_artem/0eb966e0346e954b32e36c8a276e4ea7/external/io_tweag_rules_haskell_tutorial/lib/BUILD.bazel:6:1: error executing shell command: 
    '/nix/store/ya42d72dxpby7nplsya8fj7bzj3l7529-bash/bin/bash -c bazel-out/k8-fastbuild/bin/external/io_tweag_rules_haskell_tutorial/lib/ghc_wrapper_booleans_HaskellBuildLibrary' failed: 
    Worker strategy cannot execute this HaskellBuildLibrary action, because the action has no tools
    ```
I tried to add something silly like `tools=["external/io_tweag_rules_haskell_ghc_nixpkgs/bin/ghc"]` to the `run_shell` call, but it did not change anything.

I'd appreciate any ideas on how to solve this problem of "empty tools" and suggest to Bazel to start ghc in the persistent worker mode.

**Comment by ulysses4ever** [comment]
[Wip is here](https://github.com/tweag/rules_haskell/tree/worker/haskell). Not much of progress so far.

**Comment by arturdryomov** [comment]
@ulysses4ever, hey, have you found the solution for the `because the action has no tools` issue? I see the same error in a different rule we write and it feels like a dead end.

**Comment by arturdryomov** [comment]
JFYI — the error on our side happened cause we specified `java` as a rule `executable`. Changing it to a `java_binary` target label resolved the issue. I’m still not sure why exactly it happens but oh well.

**Comment by ulysses4ever** [comment]
@arturdryomov hey! Sorry for a really late reply! (have been in crunch for the past week)

I'm glad you solved it so fast: it took me a couple of weeks to figure that out. Indeed, turning the worker script into an `sh_rule` made Bazel happy (no `no tools` any more). This was merged in #1024.

Actually, the whole issue should have been closed long ago by #1024 and 1055.
